lib/semantics/check-stop.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2019, Arm Ltd.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "check-stop.h"
+#include "semantics.h"
+#include "tools.h"
+#include "../common/Fortran.h"
+#include "../evaluate/expression.h"
+#include "../parser/parse-tree.h"
+#include <optional>
+
+void Fortran::semantics::StopChecker::Enter(
+    const Fortran::parser::StopStmt &stmt) {
+  const auto &sc{std::get<std::optional<Fortran::parser::StopCode>>(stmt.t)};
+  const auto &sle{
+      std::get<std::optional<Fortran::parser::ScalarLogicalExpr>>(stmt.t)};
+
+  if (sc.has_value()) {
+    const Fortran::parser::CharBlock &source{sc.value().v.thing.source};
+    const auto &expr{*(sc.value().v.thing.typedExpr)};
+
+    if (!(Fortran::semantics::ExprIsScalar(expr))) {
+      context_.Say(source, "Stop code must be a scalar"_err_en_US);
+    } else {
+      if (Fortran::semantics::ExprHasTypeCategory(
+              expr, Fortran::common::TypeCategory::Integer)) {
+        // C1171 default kind
+        if (!(Fortran::semantics::ExprHasTypeKind(expr,
+                context_.defaultKinds().GetDefaultKind(
+                    Fortran::common::TypeCategory::Integer)))) {
+          context_.Say(
+              source, "Integer stop code must be of default kind"_err_en_US);
+        }
+      } else if (Fortran::semantics::ExprHasTypeCategory(
+                     expr, Fortran::common::TypeCategory::Character)) {
+        // R1162 spells scalar-DEFAULT-char-expr
+        if (!(Fortran::semantics::ExprHasTypeKind(expr,
+                context_.defaultKinds().GetDefaultKind(
+                    Fortran::common::TypeCategory::Character)))) {
+          context_.Say(
+              source, "Character stop code must be of default kind"_err_en_US);
+        }
+      } else {
+        context_.Say(
+            source, "Stop code must be of INTEGER or CHARACTER type"_err_en_US);
+      }
+    }
+  }
+  if (sle.has_value()) {
```

Comment:
Expression semantics will have already enforced these constraints since they're manifest in the parse tree.

Reply:
Are you talking about the plans or actual state? I don't see it working the way you describe it...

Reply:
The type of the parse tree for the `QUIET=` specifier is a `Scalar<Logical<Expr>>`.  Expression semantics ensures that `Scalar<>` things are scalar and that `Logical<>` things are logical.

Reply:
I suspect I broke the when I created `ExprChecker`. It waits till it gets down to the `Expr` level before performing the checks, and that's too late to enforce the constraints like `Scalar<>`.

Reply:
Understood.  I'll fix it.

Reply:
As this will affect number of my test cases in stop01.f90 the way I can't predict now, I guess it's now easier to close this PR and open new when the fix is there and redundant checks in my code are gone.

Reply:
to make things easier, I've removed those checks.